File: libraries/node/src/sources/ojn/convert.ts
/**
 * Mapper from the O2Jam source model onto URC charts.
 */
import { UrcError } from "../../error.js";
import type { Chart, NoteType } from "../../model.js";
import { buildTiming, checkHoldOverlap, roundMs } from "../shared.js";
import type { OjnDifficulty, OjnFile } from "./model.js";

const MEASURE_MS = 240000;
const METER_TOLERANCE = 1e-6;
const VERSIONS = ["Easy", "Normal", "Hard"] as const;
const SCALES = [1, 10, 100, 1000, 10000, 100000, 1000000, 10000000, 100000000, 1000000000];

const TYPE_ORDER: Record<NoteType, number> = { N: 0, LS: 1, LE: 2, M: 3, F: 4 };

/**
 * Converts every difficulty of an OJN file into URC charts.
 */
export function convertOjn(file: OjnFile): Chart[] {
	return file.difficulties.map(difficulty => convertChart(file, difficulty));
}

function convertChart(file: OjnFile, difficulty: OjnDifficulty): Chart {
	const events = [...difficulty.events].sort(
		(left, right) => left.measure - right.measure || left.position - right.position
	);

	let time = 0;
	let bpm = file.bpm;
	let fraction = 1;
	let pointer = 0;
	let measure = 0;
	let meter: [number, number] = [4, 4];
	let meterDirty = false;
	const bpmPoints: Array<[number, number, number, number]> = [[0, canonicalBpm(bpm), 4, 4]];
	const anchors: number[] = [0];
	const notes: Array<[number, number, NoteType]> = [];
	const holds = new Map<number, number>();

	for (const event of events) {
		while (event.measure > measure) {
			if (fraction - pointer < 0)
				throw new UrcError("syntax", event.offset, "measure fraction cuts before the current position");
			time += (MEASURE_MS * (fraction - pointer)) / bpm;
			anchors.push(roundMs(time));
			if (meterDirty) {
				bpmPoints.push([roundMs(time), canonicalBpm(bpm), 4, 4]);
				meter = [4, 4];
				meterDirty = false;
			}
			measure++;
			fraction = 1;
			pointer = 0;
		}
		time += (MEASURE_MS * (event.position - pointer)) / bpm;
		pointer = event.position;

		if (event.channel === 0) {
			fraction = event.value;
			const approx = fractionMeter(event.value);
			if (approx !== null && (approx[0] !== meter[0] || approx[1] !== meter[1])) {
				meter = approx;
				meterDirty = true;
				bpmPoints.push([roundMs(time), canonicalBpm(bpm), meter[0], meter[1]]);
			}
		} else if (event.channel === 1) {
			bpmPoints.push([roundMs(time), canonicalBpm(event.value), meter[0], meter[1]]);
			bpm = event.value;
		} else 
			addNote(notes, holds, roundMs(time), event.channel - 2, event.kind);
		
	}
	for (const index of holds.values())
		notes[index][2] = "N";

	const noteTimes = notes.filter(([, , type]) => type !== "LE").map(([time]) => time);
	const firstNoteTime = noteTimes.length > 0 ? Math.min(...noteTimes) : 0;

	const timingPoints = buildTiming(
		bpmPoints,
		[],
		firstNoteTime,
		".ojn",
		anchors.find(time => time >= firstNoteTime) ?? null
	);

	notes.sort(
		(left, right) => left[0] - right[0] || left[1] - right[1] || TYPE_ORDER[left[2]] - TYPE_ORDER[right[2]]
	);
	const finalNotes = notes.map(([time, lane, type]) => ({
		timestampMs: time - firstNoteTime,
		lane,
		type
	}));
	checkHoldOverlap(finalNotes);

	return {
		formatVersion: { major: 1, minor: 1 },
		metadata: {
			original: "O2Jam",
			title: file.title || "Unknown",
			artist: file.artist || "Unknown",
			creator: file.noter || "Unknown",
			version: VERSIONS[difficulty.index]
		},
		judgment: null,
		layout: { keys: 7, specialKeys: 0, specialLanes: null },
		timing: timingPoints,
		notes: finalNotes
	};
}

function addNote(
	notes: Array<[number, number, NoteType]>,
	holds: Map<number, number>,
	ms: number,
	lane: number,
	kind: number
): void {
	if (kind === 3) {
		const index = holds.get(lane);
		if (index === undefined)
			return;
		holds.delete(lane);
		if (ms <= notes[index][0])
			notes[index][2] = "N";
		else
			notes.push([ms, lane, "LE"]);
		return;
	}
	if (holds.has(lane))
		return;
	if (kind === 2) {
		holds.set(lane, notes.length);
		notes.push([ms, lane, "LS"]);
	} else 
		notes.push([ms, lane, "N"]);
	
}

/**
 * Shortest decimal that round-trips the f32 BPM, widened to f64. Candidates
 * are enumerated explicitly instead of relying on formatter-specific
 * tie-breaking, so every language agrees.
 */
function canonicalBpm(value: number): number {
	for (let precision = 0; precision < SCALES.length; precision++) {
		const base = Math.trunc(value * SCALES[precision]);
		for (const candidate of [base, base + 1, base - 1]) {
			const text = decimal(candidate, precision);
			if (Math.fround(Number(text)) === value)
				return Number(text);
		}
	}
	return value;
}

function decimal(candidate: number, precision: number): string {
	const sign = candidate < 0 ? "-" : "";
	let digits = String(Math.abs(candidate));
	if (precision === 0)
		return sign + digits;
	digits = digits.padStart(precision + 1, "0");
	let text = `${sign}${digits.slice(0, digits.length - precision)}.${digits.slice(digits.length - precision)}`;
	while (text.endsWith("0"))
		text = text.slice(0, -1);
	if (text.endsWith("."))
		text = text.slice(0, -1);
	return text;
}

/** Meter (beats, noteValue) matching a measure fraction, if clean. */
function fractionMeter(value: number): [number, number] | null {
	for (let noteValue = 1; noteValue <= 64; noteValue++) {
		const beats = Math.round(value * noteValue);
		if (beats < 1)
			continue;
		if (Math.abs(value - beats / noteValue) <= METER_TOLERANCE) {
			const divisor = gcd(beats, noteValue);
			const reduced: [number, number] = [beats / divisor, noteValue / divisor];
			return reduced[0] === 1 && reduced[1] === 1 ? null : reduced;
		}
	}
	return null;
}

function gcd(a: number, b: number): number {
	while (b !== 0) {
		const rest = a % b;
		a = b;
		b = rest;
	}
	return a;
}

File: libraries/rust/src/sources/ojn/convert.rs
//! Mapper from the O2Jam source model onto URC charts.

use std::collections::HashMap;

use crate::error::{Result, UrcError};
use crate::model::{Chart, Layout, Metadata, Note, NoteType, Version};

use super::super::shared::{build_timing, check_hold_overlap, round_ms};
use super::model::{OjnDifficulty, OjnFile};

const MEASURE_MS: f64 = 240000.0;
const METER_TOLERANCE: f64 = 1e-6;
const VERSIONS: [&str; 3] = ["Easy", "Normal", "Hard"];
const SCALES: [f64; 10] = [
    1.0,
    10.0,
    100.0,
    1000.0,
    10000.0,
    100000.0,
    1000000.0,
    10000000.0,
    100000000.0,
    1000000000.0,
];

type BpmPoint = (i64, f64, u64, u64);
type UrcNote = (i64, u32, NoteType);

/// Converts every difficulty of an OJN file into URC charts.
pub fn convert_ojn(file: &OjnFile) -> Result<Vec<Chart>> {
    file.difficulties
        .iter()
        .map(|difficulty| convert_chart(file, difficulty))
        .collect()
}

fn convert_chart(file: &OjnFile, difficulty: &OjnDifficulty) -> Result<Chart> {
    let mut events = difficulty.events.clone();
    events.sort_by(|left, right| {
        left.measure
            .cmp(&right.measure)
            .then(left.position.partial_cmp(&right.position).unwrap())
    });

    let mut time = 0.0;
    let mut bpm = file.bpm;
    let mut fraction = 1.0;
    let mut pointer = 0.0;
    let mut measure = 0_i32;
    let mut meter = (4_u64, 4_u64);
    let mut meter_dirty = false;
    let mut bpm_points: Vec<BpmPoint> = vec![(0, canonical_bpm(bpm), 4, 4)];
    let mut anchors: Vec<i64> = vec![0];
    let mut notes: Vec<UrcNote> = Vec::new();
    let mut holds: HashMap<u32, usize> = HashMap::new();

    for event in &events {
        while event.measure > measure {
            if fraction - pointer < 0.0 {
                return Err(UrcError::new(
                    "syntax",
                    event.offset,
                    "measure fraction cuts before the current position",
                ));
            }
            time += (MEASURE_MS * (fraction - pointer)) / bpm;
            anchors.push(round_ms(time));
            if meter_dirty {
                bpm_points.push((round_ms(time), canonical_bpm(bpm), 4, 4));
                meter = (4, 4);
                meter_dirty = false;
            }
            measure += 1;
            fraction = 1.0;
            pointer = 0.0;
        }
        time += (MEASURE_MS * (event.position - pointer)) / bpm;
        pointer = event.position;

        if event.channel == 0 {
            fraction = event.value;
            if let Some(approx) = fraction_meter(event.value)
                && approx != meter
            {
                meter = approx;
                meter_dirty = true;
                bpm_points.push((round_ms(time), canonical_bpm(bpm), meter.0, meter.1));
            }
        } else if event.channel == 1 {
            bpm_points.push((round_ms(time), canonical_bpm(event.value), meter.0, meter.1));
            bpm = event.value;
        } else {
            add_note(
                &mut notes,
                &mut holds,
                round_ms(time),
                event.channel as u32 - 2,
                event.kind,
            );
        }
    }
    for &index in holds.values() {
        notes[index].2 = NoteType::N;
    }

    let first_note_time = notes
        .iter()
        .filter(|note| note.2 != NoteType::Le)
        .map(|note| note.0)
        .min()
        .unwrap_or(0);
    let timing_points = build_timing(
        &bpm_points,
        &[],
        first_note_time,
        ".ojn",
        anchors
            .iter()
            .copied()
            .find(|&time| time >= first_note_time),
    )?;

    let type_order = |kind: NoteType| match kind {
        NoteType::N => 0,
        NoteType::Ls => 1,
        NoteType::Le => 2,
        NoteType::M => 3,
        NoteType::F => 4,
    };
    notes.sort_by_key(|note| (note.0, note.1, type_order(note.2)));
    let final_notes: Vec<Note> = notes
        .into_iter()
        .map(|(time, lane, note_type)| Note {
            timestamp_ms: time - first_note_time,
            lane,
            note_type,
        })
        .collect();
    check_hold_overlap(&final_notes)?;

    Ok(Chart {
        format_version: Version { major: 1, minor: 1 },
        metadata: Metadata {
            original: "O2Jam".to_string(),
            title: non_empty(&file.title, "Unknown"),
            artist: non_empty(&file.artist, "Unknown"),
            creator: non_empty(&file.noter, "Unknown"),
            version: VERSIONS[difficulty.index].to_string(),
        },
        judgment: None,
        layout: Layout {
            keys: 7,
            special_keys: 0,
            special_lanes: None,
        },
        timing: timing_points,
        notes: final_notes,
    })
}

fn add_note(
    notes: &mut Vec<UrcNote>,
    holds: &mut HashMap<u32, usize>,
    ms: i64,
    lane: u32,
    kind: u8,
) {
    if kind == 3 {
        if let Some(index) = holds.remove(&lane) {
            if ms <= notes[index].0 {
                notes[index].2 = NoteType::N;
            } else {
                notes.push((ms, lane, NoteType::Le));
            }
        }
        return;
    }
    if holds.contains_key(&lane) {
        return;
    }
    if kind == 2 {
        holds.insert(lane, notes.len());
        notes.push((ms, lane, NoteType::Ls));
    } else {
        notes.push((ms, lane, NoteType::N));
    }
}

/// Shortest decimal that round-trips the f32 BPM, widened to f64. Candidates
/// are enumerated explicitly instead of relying on formatter-specific
/// tie-breaking, so every language agrees.
fn canonical_bpm(value: f64) -> f64 {
    for (precision, &scale) in SCALES.iter().enumerate() {
        let base = (value * scale).trunc() as i64;
        for candidate in [base, base + 1, base - 1] {
            let text = decimal(candidate, precision);
            if let Ok(parsed) = text.parse::<f64>()
                && parsed as f32 as f64 == value
            {
                return parsed;
            }
        }
    }
    value
}

fn decimal(candidate: i64, precision: usize) -> String {
    let sign = if candidate < 0 { "-" } else { "" };
    let digits = candidate.unsigned_abs().to_string();
    if precision == 0 {
        return format!("{sign}{digits}");
    }
    let padded = format!("{digits:0>width$}", width = precision + 1);
    let split = padded.len() - precision;
    let mut text = format!("{sign}{}.{}", &padded[..split], &padded[split..]);
    while text.ends_with('0') {
        text.pop();
    }
    if text.ends_with('.') {
        text.pop();
    }
    text
}

/// Meter (beats, note_value) matching a measure fraction, if clean.
fn fraction_meter(value: f64) -> Option<(u64, u64)> {
    for note_value in 1..=64_u64 {
        let beats = (value * note_value as f64).round();
        if beats < 1.0 {
            continue;
        }
        if (value - beats / note_value as f64).abs() <= METER_TOLERANCE {
            let divisor = gcd(beats as u64, note_value);
            let reduced = (beats as u64 / divisor, note_value / divisor);
            return if reduced == (1, 1) {
                None
            } else {
                Some(reduced)
            };
        }
    }
    None
}

fn gcd(mut a: u64, mut b: u64) -> u64 {
    while b != 0 {
        let rest = a % b;
        a = b;
        b = rest;
    }
    a
}

fn non_empty(value: &str, fallback: &str) -> String {
    if value.is_empty() {
        fallback.to_string()
    } else {
        value.to_string()
    }
}
